broadinstitute/pearl
# [JN-496] Save SendGrid activity logs

#### DESCRIPTION

SendGrid provides a maximum of 30 days of activity history. History includes information about deliverability of emails, as well as user interactions with emails (number of opens and clicks). This PR brings all of that event history into our database by adding a new background process that queries for new activity every 30 minutes. The primary reason for doing this is so we can keep track of history beyond 30 days, which will provide a long-term audit trail and make it easier to debug email issues as they come up. It's also a bit easier to query our Postgres database than deal with the Sendgrid API/UI directly.

In the future we can think about surfacing some of this information in the admin UI. It might be nice to show an "opened" column in the participant email history table so study admins can see if a user has read the email and follow up accordingly.

**PREREQ BEFORE MERGING/RELEASING**: Upgrade production SendGrid account to have 30-day activity history (dev is already upgraded)

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

Test backfilling:

* Start admin API server
* Wait ~1 minute for the scheduled process to kick off. You should see a log message such as: `Querying for SendGrid activity between 2023-09-04T00:34:58.257Z and 2023-10-04T00:34:58.257Z`
* You should eventually see a message like `Sendgrid activity query returned 504 events`
* If this process is running for the first time, it will backfill 30 days worth of events. This is the maximum amount made available to us by SendGrid. Otherwise, it will query for events that have occurred since the most recently recorded event.
* Verify that the events look good compared to what's in https://app.sendgrid.com/email_activity

Test paginated results:

* Delete everything from the `sendgrid_event` table: `delete from sendgrid_event where 1=1;`
* Reduce `PAGE_SIZE` down to 100 (don't go too low with this otherwise you'll get rate limited very quickly)
* Restart admin API server
* You should see a bunch of log messages such as `Sendgrid activity query returned the limit of 100 events. Querying for additional events between 2023-09-04T00:34:58.257Z and 2023-09-28T17:12:05.000Z`
* Verify that the events look good compared to what's in https://app.sendgrid.com/email_activity

Test new/updated activity:

* Modify the schedule + shedlock settings in ScheduledSendgridActivityService.java to query every minute or so
* Fire off a test email from the Admin UI and open an old email that was sent to you from the *dev or local* server.
* Wait about 5 minutes for the events to propagate to SendGrid
* Query the `sendgrid_event` table to see that the newly sent email is present in the DB, and the recently opened email event has also been recorded (you'll see the `opens_count` field has been incremented by 1)

In general, it's good to cross-reference events in the DB versus what's shown in https://app.sendgrid.com/email_activity
